src/lib/model.js

```diff
@@ -55,21 +55,54 @@ const processGroups = data => data.reduce(
   , []);
 const groupUsage = (dateFrom, dateTo) => query.usageByGroup(dateFrom, dateTo).then(processGroups);
 
+const hours = (name, colour) => ({
+  name,
+  data: Array.from({ length: 24 }, (_, n) => ({ count: 0, hour: n })),
+  colour
+});
+const dayCounts = (from, to) => (
+  (start, end,
+   diff = Math.abs(Math.round(moment.duration(start.startOf('day').diff(end.endOf('day'))).as('days'))),
+   mod = diff % 7,
+   weeks = Math.floor(diff / 7),
+   startDay = start.day() || 7,
+   extraWeekEndDays = mod && [6, 7].filter(n => startDay <= n && (startDay + mod) > n).length,
+   we = (weeks * 2) + extraWeekEndDays
+  ) => [diff - we, we]
+)(moment(from, DATE_FORMAT), moment(to, DATE_FORMAT));
+const processHourlyUsage = (nod, data) => {
+  const total = nod[0] + nod[1];
+  const traces = [nod[0] && hours('weekday', '#2746B9'), nod[1] && hours('weekend', '#31BB76')];
+  const average = hours('average', '#0B69D4');
+  data.forEach(d => {
+    traces[d.weekend].data[d.hour].count = (d.count / nod[d.weekend]);
+    average.data[d.hour].count += d.count;
+  });
+  return [...traces, nod[0] && nod[1] && {
+    ...average, data: average.data.map(o => ({ ...o, count: o.count / total }))
+  }].filter(e => e);
+};
+const hourlyUsage = (dateFrom, dateTo, searchGroup) => query.hourlyUsage(dateFrom, dateTo, searchGroup)
+  .then(processHourlyUsage.bind(null, dayCounts(dateFrom, dateTo || moment().endOf('day'))));
+
 const build = (dateFrom, dateTo, searchGroup, groupSearchText) => Promise.join(
   dailyUsage(dateFrom, dateTo, searchGroup),
   datasetUsage(dateFrom, dateTo),
   groupUsage(dateFrom, dateTo),
   query.searchTimePeriodByGroup(dateFrom, dateTo, searchGroup),
-  (daily, totals, groups, total) => ({
+  hourlyUsage(dateFrom, dateTo, searchGroup),
+  (daily, totals, groups, total, hourly) => ({
     from: dateFrom,
     to: dateTo,
     dates: datesInRange(dateFrom, dateTo || moment().endOf('day')),
     datasets: daily,
     groups,
     totals,
     currentGroup: groupSearchText,
-    total
+    total,
+    hourlyUsage: hourly
   })
 );
 
 module.exports = build;
+module.exports.hourlyUsage = hourlyUsage;
```

Comment:
Doesn't this override the property you set on line 93? Is that what you want?

Reply:
Line 103 even.

Except that's contained within the promise so it doesn't.

It's still a funny interface though...

Reply:
Is this line actually required? - I can't see it being used anywhere.

Reply:
This is just adding the `hourlyUsage` function to the list of things this module makes available. It doesn't override anything. It's only exported to make testing easier.

Reply:
Also, the module doesn't export a promise. It exports the function `build` (which returns a promise, i.e. only once it has been called). Because functions are still objects, they can have arbitrary fields set on them, like `hourlyUsage` here). This is a very common implementation, lots of node modules do this.

Reply:
Okay. That makes sense.

I missed that it was a function.

If it's only exported for testing purposes then it would be nice to move it out into a 'helpers' module at some point.